### PR TITLE
shell: Add popover to explain empty "User" field to "Add host" dialog

### LINKED ISSUE #14734: "Add new host" dialog is missing a tooltip.
Cockpit version: xxx
OS:
Page: "Add new host" dialog is missing a tooltip.

In the "Edit host" dialog there is a tooltip that says
>Leave  blank to connect to this machine as the currently logged in user. If  you enter a different username, that user will always be used when  connecting to this machine.

Why doesn't the same tooltip exist in the "Add a new host" dialog?

### PR DESCRIPTION
**Pull request by mvollmer**
Fixes #14734

### CODE REVIEW

**Comment by mvollmer** [comment]
There is some padding missing:

image

and the popover itself is very narrow:

image

File: pkg/lib/machine-add.html
Comment (andreasn) [inline pkg/lib/machine-add.html:33]: This ends up on a separate row for me for some reason, rather than next to the text entry as it does in your screenshot. Here is how it looks for me:
Screenshot from 2020-11-03 15-59-47

In your screenshot it lacks padding between the input field and the icon, though.

Comment (andreasn) [reply pkg/lib/machine-add.html:33]: Ignore the weird transparency of the popup in the screenshot above, that it some flicker that occurs for a milisecond when I press the Print screen button on my keyboard

**Review by andreasn — CHANGES_REQUESTED** [review]

File: pkg/lib/machine-add.html
Comment (andreasn) [reply pkg/lib/machine-add.html:33]: [Patternfly spacers](https://www.patternfly.org/v4/guidelines/spacers)

Comment (andreasn) [reply pkg/lib/machine-add.html:33]: Looking at the [Patternfly documentation](https://www.patternfly.org/v4/components/form/design-guidelines#popovers), it seems this is the recommended syntax:
```
      <button class="pf-c-form__group-label-help" aria-label="More info">
        <i class="pficon pf-icon-help" aria-hidden="true"></i>
      </button>
```

That makes the icon consistent with that use on the Networking page (Bond dialog) and the Logs page.
The aria-label would also ensure it works on a screen reader.

**Review by andreasn — CHANGES_REQUESTED** [review]

File: pkg/lib/machine-add.html
Comment (andreasn) [reply pkg/lib/machine-add.html:33]: More popover docs: https://www.patternfly.org/v4/components/popover/design-guidelines

Comment (KKoukiou) [reply pkg/lib/machine-add.html:33]: Thanks @andreasn for promoting the PF4 here ;) 
Here is the html example for that.
https://www.patternfly.org/v4/components/form/html#vertically-aligned-labels

File: pkg/lib/machine-add.html
Comment (mvollmer) [reply pkg/lib/machine-add.html:39]: Fixed.

File: pkg/lib/machine-add.html
Comment (mvollmer) [reply pkg/lib/machine-add.html:40]: I can't get them to work easily. Clicking the info button reloads the page.  The icon shows up as a missing glyph.
I propose we use the old style until we rewrite all the dialogs in React.

Comment (andreasn) [reply pkg/lib/machine-add.html:42]: Ok!

**Comment by mvollmer** [comment]
> There is some padding missing:

I have added two &nbsp;

**Review by KKoukiou — CHANGES_REQUESTED** [review]
@mvollmer I think this looks a bit too thin :+1: 
Screenshot

**Comment by mvollmer** [comment]
> I think this looks a bit too thin +1

Agreed. Personally, I don't feel like massaging this much more... so I just put the text at the top of the dialog. :-)

**Comment by KKoukiou** [comment]
> > I think this looks a bit too thin +1
> 
> Agreed. Personally, I don't feel like massaging this much more... so I just put the text at the top of the dialog. :-)
> 
> image

Since  you went that way I would rather contain it inside an info inline alert, but let's merge it as is (it's not bad IMO) and I can improve it in a followup.

**Review by KKoukiou — DISMISSED** [review]

**Comment by mvollmer** [comment]
Let's hear from @andreasn as well.

**Review by garrett — CHANGES_REQUESTED** [review]
I have issues with this.

1. Way too much text. Nobody will read this.
2. The text is describing fields that come below, but it's completely disconnected from the fields.

I have something in mind, but I'll have to have a little bit of time to mock it up for you. Meanwhile, I think these two points need to be addressed before this gets merged.

I'll post something in a bit; please wait.

**Comment by garrett** [comment]
Instead of doing a "proper mockup", I modified the PF4 form field @ https://www.patternfly.org/v4/components/form/#horizontal:

Here's a screenshot:

Screenshot_2020-12-14 PatternFly 4 • Form(3)


Obviously, the color selector shouldn't be an input field. And buttons are missing. And things aren't 100% perfect in other ways too. But the idea is that we get rid of the text and move the concepts to be contextual, with minimal text.

Just for reference (but not to be used 100%), the code snippet I used was:

```jsx
      <Form isHorizontal>
        <FormGroup label="Host" isRequired fieldId="add-host--host" helperText="Can be a hostname, IP address, alias name, or unique SSH resource">
          <TextInput
            value={value1}
            isRequired
            type="text"
            id="add-host--host"
            aria-describedby="add-host--host-helper"
            name="add-host--host"
            onChange={this.handleTextInputChange1}
          />
        </FormGroup>
        <FormGroup label="User name" fieldId="add-host--user" helperText="When empty, the current user is used">
          <TextInput
            value={value1}
            isRequired
            type="text"
            id="horizontal-form-name"
            aria-describedby="horizontal-form-name-helper"
            name="horizontal-form-name"
            onChange={this.handleTextInputChange1}
          />
        </FormGroup>
...
```

**Comment by mvollmer** [comment]
> Here's a screenshot:

Thanks!  @KKoukiou, can I interest you in implementing this? :-)

**Comment by KKoukiou** [comment]
> Instead of doing a "proper mockup", I modified the PF4 form field @ https://www.patternfly.org/v4/components/form/#horizontal:
> 
> Here's a screenshot:
> 
> Screenshot_2020-12-14 PatternFly 4 • Form(3)
> 
> Obviously, the color selector shouldn't be an input field. And buttons are missing. And things aren't 100% perfect in other ways too. But the idea is that we get rid of the text and move the concepts to be contextual, with minimal text.
> 
> Just for reference (but not to be used 100%), the code snippet I used was:
> 
> ```js-jsx
>       <Form isHorizontal>
>         <FormGroup label="Host" isRequired fieldId="add-host--host" helperText="Can be a hostname, IP address, alias name, or unique SSH resource">
>           <TextInput
>             value={value1}
>             isRequired
>             type="text"
>             id="add-host--host"
>             aria-describedby="add-host--host-helper"
>             name="add-host--host"
>             onChange={this.handleTextInputChange1}
>           />
>         </FormGroup>
>         <FormGroup label="User name" fieldId="add-host--user" helperText="When empty, the current user is used">
>           <TextInput
>             value={value1}
>             isRequired
>             type="text"
>             id="horizontal-form-name"
>             aria-describedby="horizontal-form-name-helper"
>             name="horizontal-form-name"
>             onChange={this.handleTextInputChange1}
>           />
>         </FormGroup>
> ...
> ```

We use the isRequired stars nowhere in cockpit. Let's not introduce them in an inconsistent way.

**Comment by KKoukiou** [comment]
> > Here's a screenshot:
> 
> Thanks! @KKoukiou, can I interest you in implementing this? :-)

Sure, but this page is not react, I suggest we merge this as is, and we can redesign it once we reactify it. Let's not do 10 things at once.

**Comment by garrett** [comment]
> We use the isRequired stars nowhere in cockpit. Let's not introduce them in an inconsistent way.

Sure, we can drop those stars. They were in the React on the PF page.

> Sure, but this page is not react, I suggest we merge this as is, and we can redesign it once we reactify it. Let's not do 10 things at once.

The text in this PR is still too long and not where it should be.

**Comment by mvollmer** [comment]
> Sure, but this page is not react, I suggest we merge this as is,

If we redo this, I would vote to not merge this PR at all.

**Comment by garrett** [comment]
I have a fix pending, using the current HTML, not really adding or changing much, so we can implement the design without having to port. (And then port later.)

**Comment by garrett** [comment]
Screenshots:

Default:
Screenshot_2020-12-14 Overview - garrett Rain

With values:
Screenshot_2020-12-14 Overview - garrett Rain(1)

**Comment by garrett** [comment]
I've moved my changes to stand-alone on https://github.com/cockpit-project/cockpit/pull/15044 instead. Same results, but squash-rebased.

As the other PR supersedes this one, I'm closing this for that.

**Pull request closed without merging**
